triplestore-jena-memory/src/main/java/zone/cogni/asquare/triplestore/jenamemory/LocalTdbRdfStoreService.java

```diff
@@ -1,317 +1,418 @@
   }
 
-  private void executeInLock(boolean lock, Runnable executeInLock) {
-    model.enterCriticalSection(lock);
-    try {
-      executeInLock.run();
+  private <R> R callQueryExecution(
+    final Supplier<QueryExecution> queryExecutionSupplier,
+    final JenaResultSetHandler<R> resultSetHandler,
+    final String context
+  ) {
+    if(StringUtils.hasLength(context)) {
+      log.warn("Context is not supported on this type of RdfStoreService but one was provided: {}", context);
     }
-    finally {
-      model.leaveCriticalSection();
+    return callQueryExecution(queryExecutionSupplier, resultSetHandler);
+  }
+
+  private <R> R callQueryExecution(
+    final Supplier<QueryExecution> queryExecutionSupplier, final JenaResultSetHandler<R> resultSetHandler
+  ) {
+    validateReadiness();
+    return Txn.calculateRead(dataset, () -> {
+      try (final QueryExecution queryExecution = queryExecutionSupplier.get()) {
+        return safeQuery(() -> resultSetHandler.handle(queryExecution.execSelect()), queryExecution);
+      }
+    });
+  }
+
+  private Model callQueryConstruct(final Supplier<QueryExecution> queryExecutionSupplier) {
+    validateReadiness();
+    return Txn.calculateRead(dataset, () -> {
+      try (final QueryExecution queryExecution = queryExecutionSupplier.get()) {
+        return safeQuery(queryExecution::execConstruct, queryExecution);
+      }
+    });
+  }
+
+  private void validateReadiness() {
+    if(!ready.get()) {
+      throw new InvalidRdfStoreServiceStateException();
     }
   }
 
+  private boolean validateInit(final File initFolder) {
+    boolean isValid = true;
+    final StringBuilder causesMsg = new StringBuilder();
+
+    log.info(".. .. Initialize TDB store - {} with data from: {}", tdbLocation, initFolder);
+    if(!initFolder.exists()) {
+      causesMsg.append("        * init folder does not exist\n");
+      isValid = false;
+    }
+    if(!initFolder.isDirectory()) {
+      causesMsg.append("        * init path is not a folder\n");
+      isValid = false;
+    }
+    if (!dataset.getDefaultModel().isEmpty()) {
+      causesMsg.append("        * store is not empty\n");
+      isValid = false;
+    }
+
+    // null if this abstract pathname does not denote a directory, or if an I/O error occurs.
+    if (initFolder.listFiles() == null) {
+      causesMsg.append("        * init directory is empty empty\n");
+      isValid = false;
+    }
+
+    if(!isValid) {
+      log.warn(".. .. Initialization failed for TDB store - {} because:\n{}", tdbLocation, causesMsg);
+    }
 
-  private <T> T executeInLock(boolean lock, Supplier<T> executeInLock) {
-    model.enterCriticalSection(lock);
+    return isValid;
+  }
+
+  public QueryExecution createQueryExecution(final String query) {
+    final QueryExecution queryExecution = QueryExecutionFactory.create(query, dataset);
+    queryExecution.setTimeout(firstResultTimeout, firstResultTimeUnit, overallTimeout, overallTimeUnit);
```

Comment:
a bit weird. I guess all 3 methods can be delegated so we only need to setTimeout in 1 place?

Reply:
yes, good catch, thank you! it's pushed